landscape/client/snap_utils.py

```diff
@@ -4,6 +4,28 @@
 from landscape.client.snap_http import SnapdHttpException
 
 
+def parse_assertion(headers, signature):
+    """Parse an assertion."""
+    # monkey patch pyyaml to ignore aliases e.g. *alias when loading
+    # for instance, a system-user assertion can have the following json:
+    #   {
+    #       [...]
+    #        "system-user-authority": "*",
+    #       [...]
+    #   }
+    # which after signing gets converted to yaml that looks like:
+    #   [...]
+    #   system-user-authority: *
+    #   [...]
+    # pyyaml tries to parse the * as the start of an alias leading to errors
+    yaml.Loader.fetch_alias = yaml.Loader.fetch_plain
+    yaml.Loader.yaml_implicit_resolvers.pop("*", None)
```

Comment:
This feels like it could have far-reaching side-effects, but I don't think we use yaml (or yaml aliases) in enough other places for it to matter.

Reply:
I've created a new class `PatchedYamlLoader(yaml.SafeLoader)` with the custom behavior and left `yaml.SafeLoader` untouched. 

For posterity's sake, these are the other attempts:

### Attempt 1

```python
    # load the specific module we want
    loader_spec = importlib.util.find_spec("yaml.loader")
    loader = importlib.util.module_from_spec(loader_spec)
    loader_spec.loader.exec_module(loader)
    # patch it
    loader.SafeLoader.fetch_alias = loader.SafeLoader.fetch_plain
    loader.SafeLoader.yaml_implicit_resolvers.pop("*", None)

    import sys
    # prints: 130478141495936 130478198317968
    print(id(sys.modules["yaml.loader"]), id(sys.modules["yaml"]))

    safe_loader = loader.SafeLoader(headers)
    assertion = safe_loader.get_single_data()
    safe_loader.dispose()
    assertion["signature"] = signature

    import yaml
    # prints: 130478141495936 130478198317968
    print(id(sys.modules["yaml.loader"]), id(sys.modules["yaml"]))

    # this raises the error as before (original behavior without patching)
    yaml.safe_load(headers)
```

If we do this, `yaml.safe_load(headers)` retains its original behavior but the catch is the modules in `sys.modules` both point to the same objects which is suspicious (ids are the same)

### Attempt 2 (reloading)

```python
    # patch
    yaml.SafeLoader.fetch_alias = yaml.SafeLoader.fetch_plain
    yaml.SafeLoader.yaml_implicit_resolvers.pop("*", None)

    import sys
    # prints: 123975180118560 123975179908128
    print(id(sys.modules["yaml.loader"]), id(sys.modules["yaml"]))

    assertion = yaml.load(headers, yaml.SafeLoader)
    assertion["signature"] = signature

    importlib.reload(yaml)
    # prints: 123975180118560 123975179908128
    print(id(sys.modules["yaml.loader"]), id(sys.modules["yaml"]))
    yaml.safe_load(headers)  # original behavior not recovered
```